build-system/tasks/runtime-test.js

```diff
@@ -541,96 +549,141 @@ function runTests() {
   const sectionMarker =
       (argv.saucelabs || argv.saucelabs_lite) ? 'saucelabs' : 'local';
 
-  let resolver;
-  const deferred = new Promise(resolverIn => {resolver = resolverIn;});
-  new Karma(c, function(exitCode) {
-    if (shouldCollapseSummary) {
-      console./* OK*/log('travis_fold:end:console_errors_' + sectionMarker);
-    }
-    server.emit('kill');
-    if (exitCode) {
-      log(
-          red('ERROR:'),
-          yellow('Karma test failed with exit code ' + exitCode));
+  // Run Sauce Labs tests in batches to avoid timeouts when connecting to the
+  // Sauce Labs environment.
+  let processExitCode;
+  if (argv.saucelabs || argv.saucelabs_lite) {
+    processExitCode = await runTestInBatches();
+  } else {
+    processExitCode = await createKarmaServer(c);
+  }
+
+  server.emit('kill');
+  exitCtrlcHandler(handlerProcess);
+
+  if (processExitCode != 0) {
+    log(
+        red('ERROR:'),
+        yellow('Karma test failed with exit code ' + processExitCode));
+    process.exitCode = 1;
+  }
+
+  /**
+   * Runs tests in batches
+   * @return {string} processExitCode
+   */
+  async function runTestInBatches() {
+    let batch = 1;
+    let startIndex = 0;
+    let endIndex = batchSize;
+    let processExitCode = 0;
+
+    log(green('Testing on ' + saucelabsBrowsers.length +
+      ' Sauce Lab browsers in total'));
+    while (startIndex < endIndex) {
+      const configBatch = Object.assign({}, c);
+      configBatch.browsers = saucelabsBrowsers.slice(startIndex, endIndex);
+      log(green('Batch #' + batch + ': Running tests on ' +
+        configBatch.browsers.length + ' Sauce Labs browser(s)...'));
+      const batchExitCode = await createKarmaServer(configBatch);
+      if (batchExitCode && processExitCode == 0) {
+        processExitCode = batchExitCode;
+      }
```

Comment:
There might be a bug here, since I'm seeing that the second batch isn't running.

https://travis-ci.org/ampproject/amphtml/jobs/461286041#L3122